File: app/src/main/java/com/hwonchul/movie/data/local/dao/ImageDao.java
package com.hwonchul.movie.data.local.dao;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import com.hwonchul.movie.data.local.model.ImageEntity;

import java.util.List;

import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;

@Dao
public interface ImageDao {

    @Query("SELECT * FROM " + ImageEntity.TABLE_NAME + " WHERE movie_id = :movieId")
    Flowable<List<ImageEntity>> findImagesByMovieId(int movieId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Completable insert(List<ImageEntity> entities);
}

File: app/src/main/java/com/hwonchul/movie/data/local/dao/ImageDao.kt
package com.hwonchul.movie.data.local.dao

import android.database.sqlite.SQLiteConstraintException
import androidx.room.Dao
import androidx.room.Query
import com.hwonchul.movie.base.data.dao.BaseDao
import com.hwonchul.movie.data.local.model.ImageEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface ImageDao : BaseDao<ImageEntity> {

    @Query("SELECT * FROM " + ImageEntity.TABLE_NAME + " WHERE movie_id = :movieId")
    fun findImagesByMovieId(movieId: Int): Flow<List<ImageEntity>>

    suspend fun upsert(entity: ImageEntity) {
        try {
            insert(entity)
        } catch (e: SQLiteConstraintException) {
            update(entity)
        }
    }

    suspend fun upsert(entities: List<ImageEntity>) {
        entities.forEach { entity ->
            try {
                insert(entity)
            } catch (e: SQLiteConstraintException) {
                update(entity)
            }
        }
    }
}